Fix lack of iconv on Cygwin and build performance regression on Windows

dra27 opened the pull request:
This is extracted from a larger piece of work previewed on caml-devel.

Two issues are addressed, both affecting 4.13:
1. Cygwin no longer ships the `iconv` command by default which means we either have to add a dependency on a _legacy_ package or come up with another way of translating Unicode path names to C constants
2. The code in `stdlib/StdlibModules` to sort out the casing for the modules is hideously slow in Cygwin and is performed on _every_ call to `make` on the root `Makefile` (which is a lot)

The second item can be mitigated by a more complicated `Makefile`, but it still requires spending 2 seconds performing a transformation on 9 lines of text.

This PR in isolation looks like utter madness. However, it is part of series of less mad changes which eliminate all uses of sed, awk, sh, tr, cut and ls from the build system. While these Unix utilities create simple one-liners, Unix implementations cannot even agree on how `sed -i` should work, let alone how `awk` regexps behave, and that's before we come across the vagaries of `LC_ALL` and `[A-Z]` suddenly ceasing to describe 26 characters. My argument is that we invest a considerable amount of effort maintaining a hugely complex portable C runtime, so maintaining a single C auxiliary program should not be too bad for us.

At present, `runtime/sak` performs 2 functions. In the full version it performs 10 functions - the additional 8 functions, however, do not require any additional utility functions (i.e. they use the file and string processing functions already defined in this PR).

`stdlib/StdlibModules` ends up completely eliminated by this PR. Firstly, `ocamldep` is already quite capable of deriving the link-order for the standard library, so it makes a certain measure of sense to use it. The fairly trivial `Makefile.link-order` then uses that dependency information to display the modules in link order. Secondly, by generating those lines, the C program may rely on there being no extraneous whitespace, etc.

Looking at recent runs on AppVeyor on my fork, it looks like this shaves 2-3 **minutes** off the build time (which is consistent with the ~120 calls made to make)

dra27 committed:
Use build_config.h instead of -DOCAML_STDLIB_DIR

Cygwin no longer provides the iconv command by default, which means the
translation of LIBDIR into OCAML_STDLIB_DIR cannot be guaranteed to
succeed. Technically, an unusual PREFIX could also result in an invalid
C constant in the Unix build.

Introduce a (relatively) small C program to carry out the function of
iconv and also translate the constant on the Unix side as well.

dra27 commented:
Also running through [precheck](https://ci.inria.fr/ocaml/job/precheck/601/).

xavierleroy commented:
I tried twice to review this PR and gave up.

First, I'd need some specification of what each `sak` command is supposed to do.  It could be a couple of well-chosen examples of inputs and the corresponding outputs. I have a general idea of what `sak translate` does, but no idea whatsoever for `sak prefix`.

Second, string processing in C is a complete pain (I think it's the second-worst language for this purpose, the worst being Fortran).  For example, you're using `strcpy` but you should not, because it triggers security warnings in OpenBSD and perhaps even in MSVC.  If you insist on writing `sak` in C, we need a small library of safe string operations.

Writing `sak` in a more appropriate language would avoid this chore.  Even C++ / STL would be preferable.  Python or Perl would probably shrink the code to 5 lines per `sak` function.

dra27 commented:
I must admit I'm puzzled by the volte face from the replies on 25 May on caml-devel when I introduced this work, which included my case for using C, rather than adding dependencies on alternate scripting languages/compilers?

I had assumed it would be less awkward to review `runtime/sak.c` as one, but would it be therefore be worth splitting this PR into two, with the iconv part separated from the stdlib-prefixing/build-performance issue?

Regarding the C code itself, the idea, apart from a cheeky use of our `char_os` and `*_os` definitions in `misc.h`, was that `sak.c` was self-contained. We use `strcpy` in (old) other areas without problem, in particular `ocamlyacc` without issue, the code passes precheck on all these platforms, and I had intentionally ensured that the buffers allocated for the target in the 3 calls to `strcpy` are definitely large enough. The sak binary itself is never installed, so it can't cause issues with any binary artefact tainting, so I don't follow why ocamlyacc can be installed using strcpy, but this can't?

dra27 commented:
`sak prefix` takes a file containing the raw list of basenames for the Standard Library, in link order, and emits the `STDLIB_MODULE_BASENAMES`, `STDLIB_PREFIXED_MODULES` and `STDLIB_MODULES` make variables which previous lived in `stdlib/StdlibModules`. For example:

```
dra@thor:~/ocaml/stdlib$ cat l.tmp
camlinternalFormatBasics
stdlib
stdlib__Sys

dra@thor:~/ocaml/stdlib$ ../runtime/sak prefix l.tmp
STDLIB_MODULES = \
  camlinternalFormatBasics \
  stdlib \
  stdlib__Sys
STDLIB_MODULE_BASENAMES = \
  camlinternalFormatBasics \
  stdlib \
  sys
STDLIB_PREFIXED_MODULES = \
  sys
```

`sak translate argument` emits the C string constant of argument. On Unix, that's just simple escaping (and has never been hit as an issue before), on Windows the argument is received in UTF-16/UCS-2 and is translated to an `L""` constant with escapes, for example:

```
C:\Scratch\OCaml\runtime>sak translate "C:\Program Files\Бактріан🐫"
L"C:\\Program Files\\\x0411\x0430\x043a\x0442\x0440\x0456\x0430\x043d\xd83d\xdc2b"
```

xavierleroy commented:
> I must admit I'm puzzled by the volte face from the replies on 25 May on caml-devel 

I know about this discussion.  I may have overestimated my abilities to review your code.  At any rate no one seems willing to review this PR, so we have a problem.

The issue with unsafe string functions was already pointed out in this discussion.  As to ocamlyacc, I am surprised OpenBSD doesn't complain.  At any rate, ocamlyacc is based on BSD code written 40 years ago, in what was idiomatic C code then.  Freshly-written C code should be held to higher standards.

gasche commented:
How realistic would it be to use Frama-C (... or some other verification tool, maybe [Mopsa](https://gitlab.com/mopsa/mopsa-analyzer)?) to prove the absence of out-of-bounds errors in the `sak.c` codebase? (My own experience is with Why3, not Frama-C; there is a WhyML-to-C extraction tool nowadays, but this would be getting esoteric.)

(I see both ends of the discussion: writing anything in C is unpleasant, especially scripts, but adding a dependency on another toolchain is always moderately painful.)

Octachron commented:
I am planning to review this PR in the upcoming weeks. Concerning the complexity, it seems to me that it might be simpler to reduce the build system complexity induced by the capitalization of module names by simply the standard library files.

dra27 committed:
Compute STDLIB_MODULES with a C auxiliary

Eliminates the dependency on many invocations of `tr` and `cut` to a
single invocation of a C auxiliary.

On Windows, the C auxiliary is considerably faster than the large number
of calls to tr and cut which were necessary in stdlib/StdlibModules.

dra27 commented:
Revised, slightly less C-string-processy implementation. The first commit is unaltered aside from the additional of further comments and the removal of the unneeded `#include <windows.h>`.

I've redone the prefixing part to use the command line instead of an intermediate file and, for now, restored the need to specify the link order of modules by hand, moving that part to the non-4.13 work.

That eliminates virtually all the string processing and the entirety of the file I/O in `runtime/sak`. The command line used for `sak prefix` is around the 500 character mark at the moment, so we're absolutely nowhere near the Windows limit.

The GNU make trick:

```make
FOO = $(eval FOO := $$(shell <cmd>))$(FOO)
```

used in `stdlib/StdlibModules` for the `STDLIB_MODULES` macro replaces what was being achieved with `stdlib/Makefile.stdlib_modules` in my previous version. This trick has been possible since `$(eval ` was added in 2002 to GNU make 3.80, which means it works on macOS (and I checked...). It's quite nicely documented [in this blog](https://make.mad-scientist.net/deferred-simple-variable-expansion/).

dra27 commented:
(it's been through [precheck](https://ci.inria.fr/ocaml/job/precheck/605/) again as well)

Octachron commented on runtime/Makefile:
```diff
@@ -261,6 +242,15 @@ caml/jumptbl.h : caml/instruct.h
+C_LITERAL = $(shell ./sak$(EXE) translate '$(1)')
+
+build_config.h: | sak$(EXE)
+	echo '#define OCAML_STDLIB_DIR $(call C_LITERAL,$(LIBDIR))' > $@
```
Would it make sense to append a line of comment to explain that this file is generated ?

dra27 replied:
Yes!

Octachron commented on runtime/sak.c:
```diff
@@ -0,0 +1,103 @@
+    /* On Windows, allow 7-bit printable characters to be displayed literally
+       and escape everything else (using the older \x notation for increased
+       compatibility, rather than the newer \U. */
+    } else if (c < 0x80 && iswprint(c)) {
```
Is the condtion `iswprint` necessary? The documentation said that this function is locale-sensitive, could that be an issue?

dra27 replied:
The locale part is "safe", and is the reason for rolling our own C program vs shell utilities. C _requires_ that programs start in the C locale, so in `runtime/sak` we don't have to worry about the environment settings (`sed`, `sort`, etc. call `setlocale(LC_ALL, "")` to inherit the environment's setting).

This doesn't have to be done using `iswprint` - it's equivalent to:

```suggestion
    } else if (c > 0x1F && c < 0x7F) {
```
if you'd prefer?

Octachron replied:
The function `iswprint` itself is fine. I was just wondering why escaping the control character is necessary on Windows.

dra27 replied:
Paranoia more than anything else - the lower control characters (0x00-0x1F) should never be seen because they're not valid in filenames. I just don't trust that _all_ `cl`s which are out there would correctly read 0x1F (which is valid in a filename, if tricky to type in the Windows shell) in a source string!

Octachron commented:
This looks reasonable to me. Without IO and most of the string processing, the complexity of the C code seems comparable to the complexity of the Makefile and shell plumbing it replaces.

shindere commented:
@Octachron: I don't understand your comment at https://github.com/ocaml/ocaml/pull/10451#issuecomment-872203602

What do you mean by `Concerning the complexity, it seems to me that it might be simpler to reduce the build system complexity induced by the capitalization of module names by simply the standard library files.`,
please? Would you mind elaborating a bit on this? I'm sorry, I'm unfamiliar
with the prefixing logic of the standard library. It always felt a strange
business to me, one of those I should stay away from, but looking back this
now feels stupid so I would really appreciate some catch-up course. Could
take place off-line during a coffee break, although the written form has its
advantages, too: it remains a bit longer than spoken words and others get a
chance to participate, if they aren't bored. ;-)

Octachron commented:
I meant that we could capitalize the file names (e.g. `list.ml` to `List.ml`) of the standard libraries to make the mapping between the file names and prefixed module names more straightforward . I think that this would remove some complexity in the standard library Makefile (and potentially remove a hack in the Makefile for the odoc generation).

shindere commented:
@dra27: regarding the following line in `runtime/Makefile`:

```
build_config.h: $(ROOTDIR)/Makefile.config | sak$(EXE)
```

Why do you use the `|`? It actually looks incorrect to me to use it here.

Regarding sak's `translate` command, how about a more descriptive name?

To me it looks more like an encoding than like a translation, so perhaps a
name like `encode-string` of even `encode-string-for-c` or
`encode-string-as-c-litteral`? Another potential improvement, which is not
incompatible with the previous suggestions inmy mind, would be to add a
`-h`/`--help` option to sak that would provide short descriptions of what
the commands do. For instance you could then rename `translate` to
`encode-string` and its description would be `Encode a string as a C
litteral`, perhaps with some restriction on which type of strings are taken
into account.

shindere commented:
Rather than including the examples as comments, how about turning them
into ocamltest tests? I can help with that if required.

shindere commented:
One general question on this PR: since it is OS dependent, how will
it work in the context of cross-compilation?

In such a context, would we need to deduplicate `OCAML_STDLIB_DIR` for
instance? I have to admit it's a bit unclear to me at the moment so just
writing here what comes to my mind.

shindere commented:
Same question for the second commit, in the main Makefile:

```
library: ocamlc | $(SAK)
```

I odn't understand the use of the `|`, which also looks incorrect to me
here. I don't think the tool is an order-only prefix like directories, i.e.
if the tool has been updated then everything depending on it should be
rebuilt, I think.

shindere commented:
There are also other uses of `|` in the secod commit, to which my question
also applies.

Regarding sak's prefix command, here again the name looks too generic to be
descriptive to me. How about `prefix-stdlib-module` or `prefix-module`, for
instance?

I must confess that I didn't understand what this command is doing. Does it
do something more than make's addprefix function?

shindere commented:
OK reading the code further I kind of understand, but see:

```
void prefix_stdlib_modules(int count, char_os **names)
```

Why not use the same name for the command itself and just replace
underscores by dashes?

And that, too, could perhaps be tested by ocamltest.

shindere commented:
Florian Angeletti (2021/08/16 00:55 -0700):
> I meant that we could capitalize the file names (e.g. `list.ml` to
> `List.ml`) of the standard libraries to make the mapping between the
> file names and prefixed module names more straightforward . I think
> that this would remove some complexity in the standard library
> Makefile (and potentially remove a hack in the Makefile for the odoc
> generation).

Okay! Understood now! I like the idea! It would definitely make things
more straightforward!

dra27 commented:
Thanks for the review, @shindere! Some extra commits pushed and answering your points:

- In `runtime/Makefile`, the order-only dependency on `sak$(EXE)` was intentional - `runtime/build_config.h` doesn't depend on the _executable_ (i.e. you could clean `runtime/sak$(EXE)` and still expect to build with the previous `build_config.h`). I could add a direct dependency on `sak.c`?! I don't mind the full rebuild, but I wasn't convinced it was necessarily useful.
- However, the `| $(SAK)` indirect dependencies in the root `Makefile` were a mistake which was a consequence of how I've extracted this PR from a much larger piece of work - thanks for noticing!! The larger piece of work eliminates `stdlib/StdlibModules` completely and `stdlib/Makefile.stdlib_modules` is instead generated (in the same way as `runtime/build_config.h` is in this PR now). I've removed those incorrect dependencies on `$(SAK)` and added definitions to `Makefile.common` to ensure that `$(SAK)` is definitely built if `stdlib/StdlibModules` is `include`'d (it's already a requirement that `Makefile.common` is included first). With that new commit, you can now do `make world ; make -C stdlib clean ; rm runtime/sak ; make library` and `stdlib.cma` rebuilds with the all the modules. Before that commit, you got an empty library.
- Technically speaking, `sak -help` already works (and `sak -aider`?!), since `-help` isn't a command and the utility displays its usage if it doesn't recognise the command (or if no command line is given).
- I'm not sure about having ocamltest's for it - it feels bad to add knowledge of this tool to ocamltest for it to be able to run it? The correct operation of the `translate` command is verified on each CI run (AppVeyor has always included a tortuous - in string terms - path!)
- I don't mind what the names of the commands are - I pushed a commits changing `translate` to `encode-C-literal` and `prefix` to `add-stdlib-prefix` and using the same function names in `runtime/sak.c`

johnwhitington commented:
The following is a rather tasteless suggestion, but it occurs to me each time I think about how to remove sed and awk from the OCaml build process, so I'm going to air it in public to exorcise it.

Why not vendor in Lua? It's tiny, it compiles in a few seconds, on any computer, with just C and make. It has a regular expressions library built in. It's license-compatible and stable. Its use could simplify and make safe the writing of some or many of the utilities described here.

dra27 commented:
@johnwhitington - I don't think it's tasteless, although it might be overkill! The version of [`runtime/sak.c`](https://github.com/dra27/ocaml/pull/40/files#diff-2f6e7f7bf51534114413e8638aaea9067c25cabeacf02b885e2fd9165dbf92b6) which actually eliminates _all_ sed and awk calls is only 600 lines of C, and it needs rebasing once this PR is in (there are some simplifications made here which need propagating which should hopefully make it shorter still).

shindere commented:
David Allsopp (2021/08/26 09:27 -0700):
> Thanks for the review, @shindere! Some extra commits pushed and answering your points:
>
> - In `runtime/Makefile`, the order-only dependency on `sak$(EXE)` was
> intentional - `runtime/build_config.h` doesn't depend on the
> _executable_ (i.e. you could clean `runtime/sak$(EXE)` and still
> expect to build with the previous `build_config.h`). I could add a
> direct dependency on `sak.c`?! I don't mind the full rebuild, but I
> wasn't convinced it was necessarily useful.

Useful I don't know, but let's say perhhaps more readable. Although I
don't quite understand why the dependency on the `.c` file is correct,
sorry.

> - However, the `| $(SAK)` indirect dependencies in the root `Makefile`
> - were a mistake which was a consequence of how I've extracted this PR
> - from a much larger piece of work - thanks for noticing!! The larger
> - piece of work eliminates `stdlib/StdlibModules` completely and
> - `stdlib/Makefile.stdlib_modules` is instead generated (in the same
> - way as `runtime/build_config.h` is in this PR now). I've removed
> - those incorrect dependencies on `$(SAK)` and added definitions to
> - `Makefile.common` to ensure that `$(SAK)` is definitely built if
> - `stdlib/StdlibModules` is `include`'d (it's already a requirement
> - that `Makefile.common` is included first). With that new commit, you
> - can now do `make world ; make -C stdlib clean ; rm runtime/sak ;
> - make library` and `stdlib.cma` rebuilds with the all the modules.
> - Before that commit, you got an empty library.

Okay, thanks.

> - Technically speaking, `sak -help` already works (and `sak
> -aider`?!), since `-help` isn't a command and the utility displays
> its usage if it doesn't recognise the command (or if no command line
> is given).

Yes, I noticed later, sorry for the useless comment.

> - I'm not sure about having ocamltest's for it - it feels bad to add
> knowledge of this tool to ocamltest for it to be able to run it? The
> correct operation of the `translate` command is verified on each CI
> run (AppVeyor has always included a tortuous - in string terms -
> path!)

Okay, no problem.

> - I don't mind what the names of the commands are - I pushed a commits
> - changing `translate` to `encode-C-literal` and `prefix` to
> - `add-stdlib-prefix` and using the same function names in
> - `runtime/sak.c`

Thanks. Just one question. Have you ocnsidered using Bash's `${foo^}` construct
to capitalize the first letter of a word? Couldn't that be used, in
conjunction with make's `addprefix` function to deal wiht the
prefixing of the stdlib?

dra27 commented:
> > Thanks for the review, @shindere! Some extra commits pushed and answering your points: - In `runtime/Makefile`, the order-only dependency on `sak$(EXE)` was intentional - `runtime/build_config.h` doesn't depend on the _executable_ (i.e. you could clean `runtime/sak$(EXE)` and still expect to build with the previous `build_config.h`). I could add a direct dependency on `sak.c`?! I don't mind the full rebuild, but I wasn't convinced it was necessarily useful.
>
> Useful I don't know, but let's say perhhaps more readable. Although I don't quite understand why the dependency on the `.c` file is correct, sorry.

`build_config.h` needs `sak.c` to have been compiled in order to _build_ it, but just because you recompiled the `sak` binary doesn't mean you have to rebuild `build_config.h` (same is true of "depending" on `ocamlc` or `ocamlopt`). The decision that `build_config.h` is stale is based on whether the files _consumed_ by the commands and used to make the commands have altered. I've pushed a commit making `sak$(EXE)` a normal dependency because technically it's not that it depends on `sak.c` - it also depends `caml/misc.h` and all the headers included by that which all just starts to get unnecessarily messy, for something that was very pedantically subtle in the first place...!

Are the other additional commits I pushed OK?

dra27 commented:
> Thanks. Just one question. Have you ocnsidered using Bash's `${foo^}` construct to capitalize the first letter of a word? Couldn't that be used, in conjunction with make's `addprefix` function to deal wiht the prefixing of the stdlib?

I hadn't, but this doesn't work - OCaml is built on systems which don't install bash by default (let alone Bash 4+).

shindere commented:
David Allsopp (2021/08/30 04:23 -0700):
> > Thanks. Just one question. Have you ocnsidered using Bash's `${foo^}` construct to capitalize the first letter of a word? Couldn't that be used, in conjunction with make's `addprefix` function to deal wiht the prefixing of the stdlib?
>
> I hadn't, but this doesn't work - OCaml is built on systems which
> don't install bash by default (let alone Bash 4+).

I thought our build system was requiring bahs anyway but it's definitely
not requiring anythign about bash's version.

dra27 commented:
> I thought our build system was requiring bahs anyway but it's definitely not requiring anythign about bash's version.

`${var^}` was added in Bash 4, which means the default bash in macOS doesn't have it (as with GNU make, it's stuck at 15 year old versions in order to avoid GPLv3). It's only the testsuite which requires bash (in a couple of places).

Octachron commented on runtime/sak.c:
```diff
@@ -0,0 +1,148 @@
+  if (argc == 3 && !strcmp_os(argv[1], T("encode-C-literal"))) {
+    encode_C_literal(argv[2]);
+  } else if (argc > 1 && !strcmp_os(argv[1], T("add-stdlib-prefix"))) {
+    add_stdlib_prefix(argc - 2, &argv[2]);
```
Isn't `&argv[2]` just `argv + 2`?

dra27 replied:
Yes, they're identical (including the code generated). To me the `&argv[2]` more clearly "looks" like an array with the first two elements removed than the pointer arithmetic version, but I have no strong opinion!

Octachron reviewed (approved):
This is not the most elegant code, but it does fix two build issues on Windows by removing the dependency on `iconv` and limiting the number of call to the shell in `stdlib/StdlibModules`.

dra27 commented:
One day I shall submit something elegant (and not containing `#ifdef _WIN32` somewhere...!). Commit history tidied back to commits, running through [precheck job 619](https://ci.inria.fr/ocaml/job/precheck/619/) and otherwise good to go with CI.

dra27 commented:
Oops - `s/make/$(MAKE)/` (pax BSD) - let's try [precheck job 620](https://ci.inria.fr/ocaml/job/precheck/620/) instead!
